google/cloud/bigtable/benchmarks/mutation_batcher_throughput_benchmark.cc

```diff
@@ -0,0 +1,275 @@
+            << "\n# Mutations per Batch: " << options->batch_size
+            << "\n# Concurrent Batches: " << options->max_batches << std::endl;
+
+  std::atomic<int> fails{0};
+  std::atomic<int> successes{0};
+  std::atomic<bool> timeout{false};
+
+  // Conditionally start the timeout thread
+  std::thread* timeout_thread = nullptr;
+  if (options->max_time.count() > 0) {
+    auto deadline = std::chrono::system_clock::now() + options->max_time;
+    timeout_thread = new std::thread([deadline, &timeout] {
+      while (std::chrono::system_clock::now() < deadline) {
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+      }
+      timeout = true;
+    });
+  }
+
+  // Start the MutationBatcher thread
+  CompletionQueue cq;
+  std::thread cq_runner([&cq] { cq.Run(); });
+
+  int key_width = 0;
+  for (auto i = options->mutation_count - 1; i > 0; i /= 10) ++key_width;
+  auto write = [&options, &batcher, &cq, &fails, &successes, &timeout,
+                key_width](std::int64_t const& start, std::int64_t const& end,
+                           bool const& log) {
+    // One write thread will log its progress
+    if (log) std::cout << "#\n# Writing" << std::flush;
+    auto progress_period = (std::max)(1L, (end - start) / 20);
+
+    for (std::int64_t i = start; i < end; ++i) {
+      // Stop writing if we hit the cutoff deadline
+      if (timeout) break;
+
+      auto mut = MakeMutation(*options, key_width, i);
+      auto admission_completion = batcher.AsyncApply(cq, mut);
+      auto& admission_future = admission_completion.first;
+      auto& completion_future = admission_completion.second;
+      completion_future.then([&fails, &successes](future<Status> fut) {
+        auto status = fut.get();
+        if (!status.ok())
+          ++fails;
+        else
+          ++successes;
+      });
+      admission_future.get();
+
+      if (log && (i - start) % progress_period == 0)
+        std::cout << "." << std::flush;
+    }
+    if (log) std::cout << "\n#" << std::endl;
+  };
+
+  auto timer = Timer::PerThread();
+
+  std::vector<std::future<void>> tasks;
+  std::int64_t start = 0;
+  for (int j = 0; j < options->thread_count; ++j) {
+    auto end =
+        (std::min)(options->mutation_count,
+                   start + options->mutation_count / options->thread_count);
+    tasks.emplace_back(
+        std::async(std::launch::async, write, start, end, j == 0));
+    start = end;
+  }
+
+  for (auto& t : tasks) t.get();
+
+  // Wait for all mutations to complete
+  batcher.AsyncWaitForNoPendingRequests().get();
+
+  // Join the MutationBatcher thread
+  cq.Shutdown();
+  cq_runner.join();
+
+  auto snapshot = timer.Sample();
+  std::cout << "MutationCount,BatchSize,MaxBatches,ElapsedMicroseconds,"
+               "Successes,Fails\n"
+            << options->mutation_count << "," << options->batch_size << ","
+            << options->max_batches << "," << snapshot.elapsed_time.count()
```

Comment:
The `PerThread` timer may be overkill here.  Consider just using a steady clock if you are not interested in CPU time.  Or use the `PerProcess` timer and report the CPU time?